susemanager-utils/susemanager-sls/src/_modules/uyuni_users.py

```diff
@@ -0,0 +1,825 @@
+# coding: utf-8
+from typing import Any, Dict, List, Optional, Union, Tuple
+import ssl
+import xmlrpc.client  # type: ignore
+import logging
+import datetime
+import copy
+import os
+
+import salt.config
+import salt.client
+import salt.utils.crypt
+from salt.exceptions import CommandExecutionError
+from salt.utils.minions import CkMinions
+
+log = logging.getLogger(__name__)
+
+__pillar__: Dict[str, Any] = {}
+__context__: Dict[str, Any] = {}
+__virtualname__: str = "uyuni"
+
+
+class UyuniUsersException(Exception):
+    """
+    Uyuni users Exception
+    """
+
+
+class UyuniChannelsException(Exception):
+    """
+    Uyuni channels Exception
+    """
+
+
+class RPCClient:
+    """
+    RPC Client
+    """
+    __instance__: Optional["RPCClient"] = None
+    __instance_path__: str = "/var/cache/salt/minion/uyuni.rpc.s"
+
+    def __init__(self, url: str, user: str, password: str):
+        """
+        XML-RPC client interface.
+
+        :param url: URL of the remote host
+        :param user: username for the XML-RPC endpoints
+        :param password: password credentials for the XML-RPC endpoints
+        """
+        if self.__instance__ is not None:
+            raise UyuniUsersException("Object already instantiated. Use init() method instead.")
+
+        ctx: ssl.SSLContext = ssl.create_default_context()
+        ctx.check_hostname = False
+        ctx.verify_mode = ssl.CERT_NONE
+
+        self.conn = xmlrpc.client.ServerProxy(url, context=ctx)
+        self._user: str = user
+        self._password: str = password
+        self.token: Optional[str] = None
+
+    def save_session(self) -> bool:
+        """
+        Save session of the RPC
+
+        :return: boolean, True on success
+        """
+        os.makedirs(os.path.dirname(RPCClient.__instance_path__), exist_ok=True)
+        with open(RPCClient.__instance_path__, 'wb') as fh:
+            try:
+                fh.write(str(self.get_token()).encode())
+                log.debug("Wrote RPC session to %s", RPCClient.__instance_path__)
+                ret = True
+            except OSError as exc:
+                ret = False
+                log.error("Unable to serialise RPC client: %s", exc)
+            except Exception as exc:
+               ret = False
+               log.error("Unhandled error while serialising RPC client object: %s", exc)
+
+        return ret
+
+    @staticmethod
+    def load_session() -> Optional[str]:
+        """
+        Read previously saved RPC session token from the disk.
+        If this is not possible, None is returned.
+
+        :return: string or None
+        """
+        obj: Optional[str] = None
+        try:
+            with open(RPCClient.__instance_path__, 'rb') as fh:
```

Comment:
What are the pros and cons of keeping the authentication token on disk vs. in memory?

Reply:
Why would you do that to yourself!

Reply:
> On disk!.. Why would you do that to yourself?

Sorry, this comment does not help me understand why you chose to save the token to disk instead of keeping it in memory, which is the goal of the question.

Reply:
Well, think of it. Why would you invent the whole disk writing/reading, fiddle with mutexes etc, if Salt is always running anyway and thus so you can just keep a short string in memory and refer to it from everywhere? Sure, you can use SDB or message store on the disk, polluting it with your custom module data as well, but why? It is not that you restar Salt few million times a day, so DoS'ing DB by login tokens is not the case here. As well as if you even lose those five-ten-hundred due to very frequent Salt restarts, they will still go away eventually on their timeout. Caching only makes sense for tools like `spacecmd`, which has credentials in the config and if you call it from your Shell scripts _very frequently_, you don't want it generate hundreds of bogus session IDs.

Reply:
Sorry, your comment seems to suggest it is not a good idea to store tokens on the disk, but this is what the code does:

```python
       with open(RPCClient.__instance_path__, 'wb') as fh:
            try:
                fh.write(str(self.get_token()).encode())
```

What am I missing?

Reply:
Aahh, right. Sorry, I completely messed up the context, apologies. So yeah, generally, I am totally for keeping it in memory, as I wrote above. If there could be a better way, like preserving common shared environment etc, then I would recommend to get rid of writing that to disk and store token there instead.

However, only that one is different: in _this_ case if Lazy Loader keeps flushing all the time `__context__` on running Salt, leaving no link between the session and newly loaded module, since module acquires copies of it (so storing stuff there will flush it, obviously). :disappointed: They seem changed that behaviour recently, because before it was actually still possible. But as I remember Salt had enough issues related to stale data in the context, so seems the decision was to just flush it all at once.

Storing it dynamically in current configuration instance also not working, because `__opts__` are, again, copies (there are four of them in different places, to be precise) and gets reset on Loader (and that's unfortunate, because it leads to the inconsistency inside the running Minion, but that's another issue).

Reply:
P.S. The reason why I forgot the context already, because originally it was implemented all in states modules and there you don't need to do that. Not at least today. But then I splitted state module and exec module and hit that problem.